codalab/worker/worker.py

```diff
@@ -238,10 +218,41 @@ def cleanup(self):
         logger.info("Stopped Worker. Exiting")
 
     def signal(self):
-        # When the terminate flag is False, set the stop flag to stop running
-        # the worker without changing the status of existing running bundles.
-        if not self.terminate:
+        # When the pass_down_termination flag is False, set the stop flag to stop running
+        # the worker without changing the status of existing running bundles. Otherwise,
+        # restage all bundles that are not in the terminal states [FINISHED, STAGED].
+        if not self.pass_down_termination:
             self.stop = True
+        else:
+            self.terminate_and_restage = True
+
+    def process_terminate_signal(self):
+        # If received pass_down_termination signal from CLI to terminate the worker,
+        # wait until all the existing unfinished bundles are restaged, then stop the worker.
+        if self.terminate_and_restage:
+            if self.restage_bundles() == 0:
+                self.stop = True
+
+    def restage_bundles(self):
+        restaged_bundles = []
+        terminal_stages = [RunStage.FINISHED, RunStage.STAGED]
+        for uuid in self.runs:
+            run_state = self.runs[uuid]
+            if run_state.stage not in terminal_stages:
+                self.restage(uuid)
+                restaged_bundles.append(uuid)
+        if len(restaged_bundles) == 0:
+            # reset the current runs to exclude bundles in terminal states: STAGED and FINISHED
+            self.runs = {
```

Comment:
Why wouldn't we do this in all cases?

Reply:
Do you mean why don't we do this in `process_runs` in the main loop?

Reply:
Looking at [your comment](https://github.com/codalab/codalab-worksheets/pull/2186/files#r421939322) now, I am wondering if you were asking "why do we exclude bundles in terminal stages (STAGED and FINISHED)?"